### PR TITLE
avm2: Introduce ParametersExt and use it inside bitmapdata

### PR DESCRIPTION
**Pull request by Dinnerbone**
We currently have so many different ways of pulling out parameters from the function, and almost every single one is wrong. The "correct" way to do it tends to be extremely verbose.

This introduces common methods for **correctly** getting the value of a given type from the function parameters slice, and converts BitmapData to use it as an initial proof of concept.

It's still missing more useful functions like getting native object types (DisplayObject from an arg, for example), but it's already a significant improvement.

### CODE REVIEW

File: core/src/avm2/globals/flash/display/bitmap_data.rs [hunk 1]
```diff
@@ -85,23 +86,10 @@ pub fn init<'gc>(
...
-                    .coerce_to_boolean();
-                let fill_color = if let Some(value) = args.get(3) {
-                    value.coerce_to_u32(activation)?
-                } else {
-                    0xFFFFFFFF
-                };
+                let width = args.get_u32(activation, 0)?;
+                let height = args.get_u32(activation, 1)?;
+                let transparency = args.get_bool(2);
+                let fill_color = args.get_u32(activation, 3)?;
```
[the remark this answers is not in the document]
Comment (Dinnerbone) [reply core/src/avm2/globals/flash/display/bitmap_data.rs:92]: The default is provided from AS, that's a holdover from before this was converted

**Comment by Lord-McSweeney** [comment]
Does this PR obsolete/supersede #7895?

**Comment by Dinnerbone** [comment]
I'd say no. That PR is on hold until everyone can agree on the design, and I think makes a nicer end goal, but this works for today.

**Review by Aaron1011 — APPROVED** [review]

File: core/src/avm2/parameters.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    fn try_get_object(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        index: usize,
+    ) -> Option<Object<'gc>> {
+        match self[index] {
+            Value::Null | Value::Undefined => None,
+            Value::Object(o) => Some(o),
+            primitive => Some(
+                PrimitiveObject::from_primitive(primitive, activation)
```
Comment (adrian17) [inline core/src/avm2/parameters.rs:112]: in most places we don't actually expect a primitive (matrix, color_transform, etc), so I'd consider making it a default? (just documenting it well).

On that note, I'd also add `get_value(idx)` just for a consistent api.

**Review by adrian17 — APPROVED** [review]

File: core/src/avm2/parameters.rs (hunk 2 above)
Comment (Dinnerbone) [reply core/src/avm2/parameters.rs:120]: Added get_value which is `this[index]` for now.
Good point about non-primitives, I think I'll need to consider that a bit though. Will handle it in a future PR if that's ok!

**Pull request merged**